File: exercises/ex5/demo/union.c
#include <stdio.h>

union options
{
  int option_a;
  float option_b;
};

int main() {
  union options o;

  // 1073741825
  printf("input value for option_a: ");
  scanf("%d", &o.option_a);

  printf("option_a %i\n", o.option_a);
  printf("option_b interpreted as %f\n", o.option_b);
}

File: exercises/ex5/demo/rust/src/union.rs
use scanf::scanf;

#[derive(Debug)]
enum Options {
    A(i32),
    B(f32),
}

fn main() {
    let mut i = 0;
    print!("input value for Option::A: ");
    scanf!("{}", i).unwrap();

    let o = Options::A(i);

    // println!("o.A {}", o.A);
    // println!("o.B {}", o.B);

    if let Options::A(inner) = o {
        println!("Option::A: {:?}", o);
        println!("Option::B by explict cast: {:?}", Options::B(inner as f32));
    }
}
